[WIP] Add Vimeo support

samanthaksanders opened the pull request.

samanthaksanders committed:
Start allowing vimeo embed

samanthaksanders committed:
Get vimeo play and stop function working

samanthaksanders committed:
Test if playvideo function is present before playing youtube player

samanthaksanders committed:
Allow multiple of the same video

thibaudcolas commented on example/index.html:
```diff
@@ -1,110 +1,192 @@
 <head>
     <title>Quicktube</title>
-    <link rel="stylesheet" href="../dist/quicktube.css"/>
+    <link rel="stylesheet" href="../dist/quicktube.css" />
```
This file doesn't seem to exist?

samanthaksanders replied:
Ah dist folder is ignored should I move to src folder?

thibaudcolas replied:
Yes please, then in the `build` script (https://github.com/springload/Quicktube.js/blob/master/package.json#L12), add a step that copies the file from `src` to `dist`.

thibaudcolas commented on example/index.html:
```diff
@@ -1,110 +1,192 @@
+            background-image: -webkit-linear-gradient(135deg, rgb(255, 179, 236) 0%, rgb(255, 217, 179) 100%);
+            background-image: -moz-linear-gradient(135deg, rgb(255, 179, 236) 0%, rgb(255, 217, 179) 100%);
+            background-image: -o-linear-gradient(135deg, rgb(255, 179, 236) 0%, rgb(255, 217, 179) 100%);
+            background-image: linear-gradient(135deg, rgb(255, 179, 236) 0%, rgb(255, 217, 179) 100%);
```
This (and `box-shadow` below) doesn't require a prefix since about 4 years ago, so we shouldn't hand-prefix those.

thibaudcolas committed:
Remove unnecessary browser prefixes

thibaudcolas commented on example/index.html:
```diff
@@ -1,110 +1,192 @@
+        
         body {
-            font-family: sans-serif;
+            font-family: helvetica, arial, sans-serif;
```
I have changed this to be the "system font" stack: https://css-tricks.com/snippets/css/system-font-stack/

thibaudcolas committed:
Use system font stack

thibaudcolas committed:
Move no-js check back to page head

thibaudcolas commented on example/index.html:
```diff
@@ -1,110 +1,192 @@
     </style>
-    <script>
-        document.documentElement.className = document.documentElement.className.replace("no-js", "js");
-    </script>
```
This is meant to stay in the `head` so the content isn't displayed with the `no-js` version first while JS is loading.

thibaudcolas committed:
Add empty alt attributes to images

thibaudcolas commented on example/index.html:
```diff
@@ -1,110 +1,192 @@
+                    <div class="quicktube__video" data-quicktube-video></div>
+                    <div data-quicktube-play="76979871" class="quicktube__poster" data-quicktube-poster>
+                        <!-- Optional poster frame image -->
+                        <!--<img src="http://www.lovethispic.com/uploaded_images/90725-Colorful-Kittens.jpg?w=240" class="quicktube__poster-image">-->
```
Even if this is just a small demo, please _always_ follow our accessibility (and other) best practices to the letter – and in this case add an empty `alt` attribute (or one with example content). Build your demos as if your code was then going to be copy/pasted as-is all around the internet.

thibaudcolas commented on example/index.html:
```diff
@@ -1,110 +1,192 @@
+                    <div class="quicktube__video" data-quicktube-video></div>
+                    <div data-quicktube-play="76979871" class="quicktube__poster" data-quicktube-poster>
+                        <!-- Optional poster frame image -->
+                        <!--<img src="http://www.lovethispic.com/uploaded_images/90725-Colorful-Kittens.jpg?w=240" class="quicktube__poster-image">-->
```
I think there should be at least one demo with a poster image.

thibaudcolas commented on example/index.html:
```diff
@@ -1,110 +1,192 @@
+    <script>
+        // Remember to initialise QuickTube at some point in your app
+        document.addEventListener('DOMContentLoaded', function() {
+            const showingItem = document.querySelector('[data-show-while-playing]');
```
Not sure why this is defined and then not used?

samanthaksanders replied:
Neither, I shall remove

thibaudcolas commented on src/quicktube.js:
```diff
@@ -21,89 +26,113 @@ const trackEvent = (event) => {
+    let url = `${playerEmbedUrl}${playerId}?autoplay=1`;
+    const optionKeys = Object.keys(options);
+    optionKeys.forEach((key) => {
+        url += `&${key}=${options[key]}`;
```
This should do `encodeURIComponent` on both key and value to make sure they are URL-friendly.

thibaudcolas committed:
Add encodeURIComponent calls to createPlayerUrl

thibaudcolas commented on src/quicktube.js:
```diff
@@ -21,89 +26,113 @@ const trackEvent = (event) => {
 };
 
-const getCurrentSegment = (currentPosition, duration) => {
+const createPlayerUrl = (playerEmbedUrl, playerId, options) => {
```
Getting very finicky here – this should be `createPlayerURL`. camelCase mandates writing acronyms as uppercase, unless they are the first part of the name.

thibaudcolas commented on src/quicktube.js:
```diff
@@ -21,89 +26,113 @@ const trackEvent = (event) => {
     const percentage = (currentPosition / duration);
+    // Ensure value is rounded to nearest whole segment eg. 1, 2, 3 , 4
     return (Math.floor(percentage * numberOfSegments) / numberOfSegments).toFixed(2);
 };
```
We're starting to get quite a few of those functions in here, so I think we should start to split this code into separate files.

samanthaksanders replied:
Not sure what exactly you want me to split when you say "we're starting to get a few"

thibaudcolas commented on src/quicktube.js:
```diff
@@ -21,89 +26,113 @@ const trackEvent = (event) => {
+    };
+
+    return `${s4()}${s4()}-${s4()}-${s4()}-${s4()}-${s4()}${s4()}${s4()}`;
+};
```
Hmm, we should use a ready-made npm package for this instead of reinventing our own. Or increment a counter instead?

thibaudcolas committed:
Change casing in playerUrl

thibaudcolas commented on src/quicktube.js:
```diff
@@ -21,89 +26,113 @@ const trackEvent = (event) => {
+    constructor(videoId, videoGUID, videoEmbedUrl, options = {}) {
+        this.videoId = videoId;
+        this.videoGUID = videoGUID;
+        this.videoEl = document.querySelector(`[data-quicktube-quid="${this.videoGUID}"]`);
```
I'm not sure I like this approach, but I'll have to read through most of the code to see if I can find one I like better 😁 .

thibaudcolas commented on src/quicktube.js:
```diff
@@ -21,89 +26,113 @@ const trackEvent = (event) => {
+        this.onPlayerError = this.onPlayerError.bind(this);
+
+        // Booleans
+        this.isMobileSafari = isMobileSafari();
```
The `isMobileSafari` check is relatively inexpensive, so let's remove this attribute and directly use the function where necessary instead.

thibaudcolas commented on src/quicktube.js:
```diff
@@ -21,89 +26,113 @@ const trackEvent = (event) => {
+
+        // Booleans
+        this.isMobileSafari = isMobileSafari();
+        this.isVimeo = this.videoEl.hasAttribute('data-quicktube-vimeo');
```
If we wanted to extend this player to support other platforms in the future, we would have to add more `data-quicktube-*` attributes. This would become hard to keep track of, so instead I would suggest having a single `data-quicktube-platform` attribute that can be equal to `youtube`, `vimeo`, etc.

thibaudcolas committed:
Inline isMobileSafari checks

thibaudcolas committed:
Introduce new data-quicktube-platform attribute

thibaudcolas commented on src/quicktube.js:
```diff
@@ -21,89 +26,113 @@ const trackEvent = (event) => {
+        this.isVimeo = this.videoEl.hasAttribute('data-quicktube-vimeo');
+
+        // Settings
         this.options = Object.assign({
```
`Object.assign` needs to be polyfilled for IE11 (see https://kangax.github.io/compat-table/es6/#test-Object_static_methods), so we will need to mention this in the project's README as well as add it on the demo page. See #6.

thibaudcolas commented on src/quicktube.js:
```diff
@@ -21,89 +26,113 @@ const trackEvent = (event) => {
             posterFrameHiddenClass: 'quicktube__poster--hidden',
+        }, options);
+
+        const playerOptions = !this.isVimeo ? {
```
Since this is a negative check, I would suggest turning it around to make this more readable. This would become:

```js
const playerOptions = this.isVimeo ? {
    autopause: 0,
} : {
    showInfo: 0,
    autohide: 1,
    color: 'white',
    playerapi: 'ytplayer',
    enablejsapi: 1,
    wmode: 'transparent',
};
```

thibaudcolas committed:
Turn ternary operator around

thibaudcolas commented on src/quicktube.js:
```diff
@@ -21,89 +26,113 @@ const trackEvent = (event) => {
+        // Need to have unique id's so that multiple of the same video can exist on a page without breaking
+        const playEl = this.videoEl.querySelector('[data-quicktube-play]');
+        playEl.setAttribute('data-play-guid', this.videoGUID);
+        const uniquePlayButton = document.querySelector(`[data-play-guid="${this.videoGUID}"]`);
```
According to this code `playEl` is equal to `uniquePlayButton`. We shouldn't have two variables pointing to the same element.

thibaudcolas committed:
Remove extra playButton variable

thibaudcolas committed:
Refactor code checking whether player is loaded already

thibaudcolas commented on src/quicktube.js:
```diff
@@ -21,89 +26,113 @@ const trackEvent = (event) => {
             iframe = videoIframes[0];
         }
 
         if (!iframe) {
```
I know this isn't part of this PR but this code needs refactoring. There is no point in using `getElementsByTagName` and having this extra `if` since all we do is check whether there is at least one `iframe` inside of the container.

Replacing this with:

```js
const iframeContainer = this.videoEl.querySelector('[data-quicktube-video]');
const hasPlayer = !!iframeContainer.querySelector('iframe');

if (!hasPlayer) {...}
```

thibaudcolas commented on src/quicktube.js:
```diff
@@ -21,89 +26,113 @@ const trackEvent = (event) => {
-                this.quicktubePlayer.playVideo();
+            if (this.quicktubePlayer) {
+                if (this.isVimeo) {
+                    this.quicktubePlayer.play();
```
This is inside the `isMobileSafari` check, although that check is about a YouTube bug. Should the `isMobileSafari`check be moved inside of the `else`?

thibaudcolas commented on src/quicktube.js:
```diff
@@ -130,28 +159,67 @@ class Quicktube {
+            this.quicktubePlayer = new Vimeo.Player(this.videoGUID);
+
+            this.quicktubePlayer.on('play', () => {
+                console.log(this.videoGUID, ': Vimeo played!');
```
This logging should only be defined in the example, not in the library itself.

thibaudcolas commented on src/quicktube.js:
```diff
@@ -130,28 +159,67 @@ class Quicktube {
+            });
+
+            // Might wanna check this functionality, may want to leave stopped player
+            // state so user can nav to other related videos?
```
Good point, this sounds like overstepping.

thibaudcolas commented on src/quicktube.js:
```diff
@@ -2,16 +2,21 @@ const KEY_CODES = {
+const YOUTUBE_EMBED = 'https://www.youtube.com/embed/';
+const VIMEO_API = 'https://player.vimeo.com/api/player.js';
+const VIMEO_EMBED = 'https://player.vimeo.com/video/';
+const FIRST_SCRIPT_TAG = document.getElementsByTagName('script')[0];
```
Prefer `querySelector` and `querySelectorAll` above all other selector methods.

thibaudcolas committed:
Use querySelector to find DOM element

thibaudcolas commented on src/quicktube.js:
```diff
@@ -264,27 +345,38 @@ class Quicktube {
+        let videoDomain;
+        if (video.hasAttribute('data-quicktube-vimeo')) {
+            // Inject the Vimeo Player API
+            insertApiScript(VIMEO_API, window.Vimeo);
```
This will lead to the player script being added multiple times, because the `Vimeo` or `YT` globals are only loaded once the script is loaded – not instantaneously when the script is injected onto the page. Instead, this function can check whether there already is a script on the page with this exact same URL.

```js
const isAlreadyInserted = document.querySelector(`[src="${url}"]`);
```

thibaudcolas committed:
Rename insertApiScript function to insertScript

thibaudcolas committed:
Change script insertion to rely on URL instead of global for duplicate detection

thibaudcolas commented on src/quicktube.js:
```diff
@@ -264,27 +345,38 @@ class Quicktube {
-        const firstDocumentScript = document.getElementsByTagName('script')[0];
-        firstDocumentScript.parentNode.insertBefore(newScriptTag, firstDocumentScript);
+        newScriptTag.src = url;
+        FIRST_SCRIPT_TAG.parentNode.insertBefore(newScriptTag, FIRST_SCRIPT_TAG);
```
It is premature optimisation to have `FIRST_SCRIPT_TAG` extracted as a global variable, and it shouldn't be uppercased – it's a mutable DOM element, not a real constant.

thibaudcolas committed:
Refactor API script loading

thibaudcolas commented on src/quicktube.js:
```diff
@@ -264,27 +345,38 @@ class Quicktube {
 // https://developers.google.com/youtube/iframe_api_reference#Requirements
 window.onYouTubeIframeAPIReady = () => {
     // TODO investigate whether this is a set requirement
 };
```
Removing this for now. The requirement is that we should only use the `YT` global once this function has been called – in our code, we will run into problems if the user clicks on the "Play" button before the API script is done loading.

Since we `init` on page load, we haven't run into this problem so far. I'll add a separate issue to fix this (#7).

thibaudcolas committed:
Replace video GUID with simpler player ID

thibaudcolas committed:
Replace separate init function with constructor function directly on class

thibaudcolas committed:
Start cleaning up the tests

thibaudcolas commented on src/quicktube.js:
```diff
@@ -264,27 +345,38 @@ class Quicktube {
         const options = JSON.parse(video.getAttribute('data-quicktube-options'));
-        const player = new Quicktube(videoId, options);
+        const videoGUID = guid();
+        video.setAttribute('data-quicktube-quid', videoGUID);
```
I have replaced this with a simpler unique player ID generation. The only place I could find it being used (apart from the logs) is `playEl.setAttribute('data-play-guid', this.playerId);`, but this attribute doesn't seem to be used anywhere else so I'm not sure it's even necessary.

thibaudcolas reviewed (approved):
@samanthaksanders here is my review. This seems ready to go so I've done lots of the changes I mention directly to speed things up – apart from the `console.log`s, I think this can be merged as-is.

I haven't had the time to look at the tests properly, will get back to this another time.

If there is one thing I would look at before merging this, it's this bit here:

```js
// Only trigger force video play if not Mobile safari as playVideo function not supported
        if (!isMobileSafari()) {
            if (this.quicktubePlayer) {
                if (this.isVimeo) {
                    this.quicktubePlayer.play();
                } else {
                    // It doesn't have playVideo function in the initial state. Is added after video is ready
                    const isLoaded = this.quicktubePlayer.playVideo;

                    if (isLoaded) {
                        this.quicktubePlayer.playVideo();
                    }
                }
            }
        }
```

... where we are restricting our use of the Vimeo API based on an issue that presumably only occurs with the YouTube player.

Let me know if you need me to work on this further over the next few days.

samanthaksanders committed:
remove redundat const from example page

samanthaksanders committed:
Add copy task to add css to dist folder on build

samanthaksanders committed:
Pull example over to our public package template style

samanthaksanders committed:
Work on refactoring some of the analytics tracking and example code

Make player state tracking functions more generic

rebase

Attempt to add percentage tracking to video progress. Not working so far need help

refactores code and fix issues with timeupdate

Make player state tracking functions more generic

refactores code and fix issues with timeupdate

Remove analytics

thibaudcolas commented on example/index.html:
```diff
@@ -11,6 +11,16 @@
+    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
+    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
+    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
+    })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
```
Please use the "alternative async snippet" available on https://developers.google.com/analytics/devguides/collection/analyticsjs/.

thibaudcolas commented on example/index.html:
```diff
@@ -11,6 +11,16 @@
+    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
+    })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
+
+    ga('create', 'UA-40614522-2', 'auto');
```
This should use a dedicated property (under the springload.github.io account?). We don't want our debugging property IDs to be available publicly.

thibaudcolas commented on src/quicktube.js:
```diff
@@ -159,6 +160,50 @@ class Quicktube {
+
+        if (this.options.trackAnalytics) {
+            // Report % played every second
+            setTimeout(this.onPlayerPercent.bind(this, target), 1000);
```
Every second feels overly wasteful. Please use a higher time (10s?), and-or make this configurable

thibaudcolas replied:
And why even have this? Wouldn't the `onPlayerPercent` tracking be enough?

samanthaksanders committed:
Add constant for tracking timeout variable and make that time length a longer interval

samanthaksanders committed:
Remove unused stop video functionality and update tests

samanthaksanders committed:
Remove unused stop test

samanthaksanders committed:
Remove any refs to lastP variable in tests since no longer used in code

samanthaksanders committed:
Move play button position styles to demo only

samanthaksanders committed:
Add option for stopping video on player end

samanthaksanders committed:
Slow down hide video transition when video ends

samanthaksanders committed:
Add error tracking function

thibaudcolas committed:
Add webpack-dev-server to facilitate development

samanthaksanders committed:
Add page title not video title to tracking if there is player error

samanthaksanders committed:
Update placeholder images and error tracking labels

thibaudcolas committed:
Add YouTube/Vimeo API links to README
